### ISSUE TITLE
Force TypeScript definition of not primitive props

### ISSUE THREAD

**Issue by przemyslawjanpietrzak** [vuejs/eslint-plugin-vue#1655]
**Please describe what the rule should do:**
Rule should throw error/warning when component has declared prop without TypeScript definition

**What category should the rule belong to?**
[ ] Enforces code style (layout)
[ ] Warns about a potential error (problem)
[ ] Suggests an alternate way of doing something (suggestion)
[x] Other: Increase type coverage

**Provide 2-3 code examples that this rule should warn about:**

```vue
<script lang="ts">
export default {
  props: {
    total: {
      type: Object, // Error
      required: true,
    },
  },
};
</script>
```

```vue
<script lang="ts">
export default {
  props: {
    total: {
      type: Object as PropType<Price>, // OK
      required: true,
    },
  },
};
</script>
```

**Additional context**

Since `defineComponent` from `vue-next` or `@vue/composition-api` TypeScript compiler understands how properties of `this` are typed. However there's no way to force writing type definition of props (other than `noImpliciteThis` in the `tsconfig`). Rule like this can improve TypeScript migrations.

Please let me know what you think about it. If you like this idea I can implement this rule.

**Comment by tatarysh** [comment]
@przemyslawjanpietrzak any news ?

**Comment by przemyslawjanpietrzak** [comment]
@tatarysh I have an implementation here: 
- https://raw.githubusercontent.com/przemyslawjanpietrzak/eslint-plugin-vue-extras/master/lib/rules/type-object-props.js
- https://raw.githubusercontent.com/przemyslawjanpietrzak/eslint-plugin-vue-extras/master/tests/lib/rules/type-object-props.js

If you need this, the simplest way will be just add to as private eslint plugin

**Comment by FloEdelmann** [comment]
@przemyslawjanpietrzak Thanks for sharing! Feel free to submit a pull request to add this to the official eslint-plugin-vue!

**Comment by przemyslawjanpietrzak** [comment]
@FloEdelmann Thanks. I'm starting the implementation of PR
